accounts-db/src/accounts.rs

```diff
@@ -572,51 +572,68 @@ impl Accounts {
+            validate_account_locks(tx.account_keys(), tx_account_lock_limit)
+                .map(|_| TransactionAccountLocksIterator::new(tx))
+        });
+        self.lock_accounts_inner(tx_account_locks_results, relax_intrabatch_account_locks)
     }
 
+    /// This function will prevent multiple threads from modifying the same account state at the
+    /// same time, possibly excluding transactions based on prior results
     #[must_use]
     pub fn lock_accounts_with_results<'a>(
         &self,
         txs: impl Iterator<Item = &'a (impl SVMMessage + 'a)>,
         results: impl Iterator<Item = Result<()>>,
         tx_account_lock_limit: usize,
+        relax_intrabatch_account_locks: bool,
     ) -> Vec<Result<()>> {
         // Validate the account locks, then get iterator if successful validation.
-        let tx_account_locks_results: Vec<Result<_>> = txs
-            .zip(results)
-            .map(|(tx, result)| match result {
-                Ok(()) => validate_account_locks(tx.account_keys(), tx_account_lock_limit)
-                    .map(|_| TransactionAccountLocksIterator::new(tx)),
-                Err(err) => Err(err),
-            })
-            .collect();
-        self.lock_accounts_inner(tx_account_locks_results)
+        let tx_account_locks_results = txs.zip(results).map(|(tx, result)| match result {
+            Ok(()) => validate_account_locks(tx.account_keys(), tx_account_lock_limit)
+                .map(|_| TransactionAccountLocksIterator::new(tx)),
+            Err(err) => Err(err),
+        });
+        self.lock_accounts_inner(tx_account_locks_results, relax_intrabatch_account_locks)
     }
 
     #[must_use]
-    fn lock_accounts_inner(
+    fn lock_accounts_inner<'a>(
         &self,
-        tx_account_locks_results: Vec<Result<TransactionAccountLocksIterator<impl SVMMessage>>>,
+        tx_account_locks_results: impl Iterator<
+            Item = Result<TransactionAccountLocksIterator<'a, impl SVMMessage + 'a>>,
+        >,
+        relax_intrabatch_account_locks: bool,
     ) -> Vec<Result<()>> {
         let account_locks = &mut self.account_locks.lock().unwrap();
-        tx_account_locks_results
-            .into_iter()
-            .map(|tx_account_locks_result| match tx_account_locks_result {
-                Ok(tx_account_locks) => {
-                    account_locks.try_lock_accounts(tx_account_locks.accounts_with_is_writable())
-                }
-                Err(err) => Err(err),
-            })
-            .collect()
+        if relax_intrabatch_account_locks {
+            let validated_batch_keys = tx_account_locks_results
+                .map(|tx_account_locks_result| {
+                    tx_account_locks_result.map(|tx_account_locks| {
+                        tx_account_locks
+                            .accounts_with_is_writable()
+                            .collect::<Vec<_>>()
+                    })
+                })
+                .collect::<Vec<_>>();
```

Comment:
This is all quite expensive and could all be done before acquiring the accounts lock.